### PR TITLE
worker: support pause in tests

### PR DESCRIPTION
**Pull request by BusyJay**
So that we are able to control async call in some way.

@siddontang @disksing @hhkbp2 PTAL

### CODE REVIEW

File: src/util/worker/mod.rs [hunk 1]
```diff
@@ -80,21 +82,53 @@ impl<T: Display, R: Runnable<T>> BatchRunnable<T> for R {
...
+#[allow(mutex_atomic)]
+impl ScheduleContext {
+    fn new() -> ScheduleContext {
+        ScheduleContext {
+            counter: Arc::new(AtomicUsize::new(0)),
+            pause: Arc::new((Mutex::new(false), Condvar::new())),
+        }
+    }
+
+    fn pause_if_neccessary(&self) {
```
Comment (siddontang) [inline src/util/worker/mod.rs]: try_pause

**Comment by siddontang** [comment]
Rest LGTM

File: src/util/worker/mod.rs [hunk 2]
```diff
@@ -80,21 +82,53 @@ impl<T: Display, R: Runnable<T>> BatchRunnable<T> for R {
     }
 }
 
+#[derive(Clone)]
+struct ScheduleContext {
+    counter: Arc<AtomicUsize>,
+    #[cfg(test)]
```
Comment (hhkbp2) [inline src/util/worker/mod.rs]: Why to add attribute in this place?

File: src/util/worker/mod.rs [hunk 3]
```diff
@@ -107,22 +141,38 @@ impl<T: Display> Scheduler<T> {
...
     /// Check if underlying worker can't handle task immediately.
     pub fn is_busy(&self) -> bool {
-        self.counter.load(Ordering::SeqCst) > 0
+        self.ctx.counter.load(Ordering::SeqCst) > 0
+    }
+
+    #[cfg(test)]
+    pub fn pause(&self) {
+        let &(ref lock, ref cond) = &*self.ctx.pause;
+        let mut guard = lock.lock().unwrap();
```
Comment (hhkbp2) [inline src/util/worker/mod.rs]: s/guard/pause/
Use consistent name as in `pause_if_neccessary()`.

**Comment by hhkbp2** [comment]
Rest LGTM

File: src/util/worker/mod.rs (hunk 1 above)
Comment (BusyJay) [inline src/util/worker/mod.rs]: I think `try_pause` should return a result or bool to indicate if it's paused. Seem not suitable here.

**Pull request closed without merging**